src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -676,70 +707,87 @@ public void Lock(string entityName, object obj, LockMode lockMode)
 			}
 		}
 
-		/// <summary>
-		///
-		/// </summary>
-		/// <param name="collection"></param>
-		/// <param name="queryString"></param>
-		/// <returns></returns>
 		public IQuery CreateFilter(object collection, string queryString)
 		{
 			using (new SessionIdLoggingContext(SessionId))
 			{
 				CheckAndUpdateSessionStatus();
 
-				CollectionFilterImpl filter =
-					new CollectionFilterImpl(queryString, collection, this,
-											 GetFilterQueryPlan(collection, queryString, null, false).ParameterMetadata);
+				var plan = GetFilterQueryPlan(collection, queryString, null, false);
+				var filter = new CollectionFilterImpl(queryString, collection, this, plan.ParameterMetadata);
 				//filter.SetComment(queryString);
 				return filter;
 			}
 		}
 
-		private FilterQueryPlan GetFilterQueryPlan(object collection, string filter, QueryParameters parameters, bool shallow)
+		public override IQuery CreateFilter(object collection, IQueryExpression queryExpression)
+		{
+			using (new SessionIdLoggingContext(SessionId))
+			{
+				CheckAndUpdateSessionStatus();
+
+				var plan = GetFilterQueryPlan(collection, queryExpression, null, false);
+				var filter = new ExpressionFilterImpl(plan.QueryExpression, collection, this, plan.ParameterMetadata);
+				return filter;
+			}
+		}
+
+		private IQueryExpressionPlan GetFilterQueryPlan(object collection, IQueryExpression queryExpression, QueryParameters parameters, bool shallow)
+		{
+			return GetFilterQueryPlan(collection, parameters,
+				role => Factory.QueryPlanCache.GetFilterQueryPlan(queryExpression, role, shallow, EnabledFilters));
+		}
+
+		private IQueryExpressionPlan GetFilterQueryPlan(object collection, string filter, QueryParameters parameters, bool shallow)
+		{
+			return GetFilterQueryPlan(collection, parameters,
+				role => Factory.QueryPlanCache.GetFilterQueryPlan(filter, role, shallow, EnabledFilters));
+		}
+
+		private IQueryExpressionPlan GetFilterQueryPlan(object collection, QueryParameters parameters, Func<string, IQueryExpressionPlan> getPlanForRole)
```

Comment:
I usually avoid this kind of pattern, putting code in common thanks to function pointers passed around. But here, with the function being defined for all cases right above, I think it is ok.

Reply:
Does not seem compatible with Async generator, causes the removal of `GetFilterQueryPlanAsync`. Now I wonder what IO are there in that function, going to check.